include/point_one/polaris_asio_embedded_client.h

```diff
@@ -0,0 +1,657 @@
+      std::istream response_stream(&response);
+      std::string http_version;
+      response_stream >> http_version;
+      unsigned int status_code;
+      response_stream >> status_code;
+      std::string status_message;
+      std::getline(response_stream, status_message);
+      if (!response_stream || http_version.substr(0, 5) != "HTTP/") {
+        LOG(ERROR) << "Invalid HTTP response to challenge request.";
+        return "";
+      }
+
+      switch(status_code) {
+        case 200:
+          VLOG(1) << "Received polaris challenge.";
+          break;
+        default:
+          LOG(ERROR) << "Challenge resquest returned with status code "
+                     << status_code << ".";
+          return "";
+      }
+
+      // Process the response headers.
+      boost::asio::read_until(socket, response, "\r\n\r\n");
+      std::string header;
+      VLOG(3) << "Response:";
+      while (std::getline(response_stream, header) && header != "\r") {
+        VLOG(3) << header;
+      }
+
+      // Get the entire reply from the socket.
+      std::stringstream reply_body;
+      if (response.size() > 0) {
+        reply_body << &response;
+      }
+      boost::system::error_code error;
+      while (boost::asio::read(socket, response,
+                               boost::asio::transfer_at_least(1), error)) {
+        reply_body << &response;
+      }
+      if (error != boost::asio::error::eof) {
+        LOG(ERROR) << "Error occured while receiving challenge: "
+                   << error.message();
+        return "";
+      }
+
+      std::string challenge = "";
+      try {
+        // Parse json.
+        boost::property_tree::ptree pt;
+        boost::property_tree::read_json(reply_body, pt);
+        challenge = pt.get<std::string>("challenge");
+        VLOG(3) << "Challenge is : " << challenge;
+      } catch (std::exception &e) {
+        LOG(ERROR) << "Exception in parsing challenge request response: " << e.what();
+        return "";
+      }
+
+      return challenge;
+    } catch (std::exception &e) {
+      LOG(ERROR) << "Exception: " << e.what();
+    }
+    return "";
+  }
+
+  std::string BuildChallengeSignature(std::string challenge, uint64_t time) {
+    // concatenate the string and calculate the message digest
+
+    // challenge| timestamp | company_id | serial_number
+    std::stringstream ss;
+    ss << challenge << std::to_string(time) << company_id_ << unique_id_;
+
+    unsigned int HMAC_SHA512_HASH_LEN = 64;
+    uint8_t buf[HMAC_SHA512_HASH_LEN];
+    std::string message = ss.str();
+    HMAC_CTX hmac_ctx;
+    HMAC_CTX_init(&hmac_ctx);
+    HMAC_Init_ex(&hmac_ctx, (void *)signing_secret_.data(),
+                 signing_secret_.length(), EVP_sha512(), NULL);
+    HMAC_Update(&hmac_ctx, (uint8_t *)message.data(), message.length());
+    HMAC_Final(&hmac_ctx, buf, &HMAC_SHA512_HASH_LEN);
+    unsigned char encodedData[128] = {};  // this is beyond safe, the max length
```

Comment:
Maybe move the blank line below to above this line to separate HMAC stuff from EVP stuff